Zend/zend_inheritance.c

```diff
@@ -173,72 +174,205 @@ static zend_always_inline zend_bool zend_iterable_compatibility_check(zend_arg_i
-		} else {
-			zend_string_addref(fe_class_name);
-		}
+		EMPTY_SWITCH_DEFAULT_CASE();
+	}
+}
+
+/* This int return is not a boolean, but will often be treated this way.
+ *   *  0 means there is definitely an error.
+ *   *  1 means there are definitely not any errors.
+ *   * -1 means there are no known errors but it is not known to be good
+ *     either (there is not enough information at the moment).
+ */
+static
+int _check_inherited_arg_info(
+	const zend_function *fe, zend_arg_info *fe_arg_info,
+	const zend_function *proto, zend_arg_info *proto_arg_info,
+	_variance variance) /* {{{ */
+{
+	zend_type fe_type = fe_arg_info->type;
+	zend_type proto_type = proto_arg_info->type;
+	zend_long fe_type_code = ZEND_TYPE_CODE(fe_type);
+	zend_long proto_type_code = ZEND_TYPE_CODE(proto_type);
+
+	ZEND_ASSERT(ZEND_TYPE_IS_SET(fe_type));
+	ZEND_ASSERT(ZEND_TYPE_IS_SET(proto_type));
+
+	if (variance == COVARIANT && ZEND_TYPE_ALLOW_NULL(fe_type) && !ZEND_TYPE_ALLOW_NULL(proto_type)) {
+		return 0;
+	}
+
+	/* This introduces BC break described at https://bugs.php.net/bug.php?id=72119 */
+	if (variance == CONTRAVARIANT && ZEND_TYPE_ALLOW_NULL(proto_type) && !ZEND_TYPE_ALLOW_NULL(fe_type)) {
+		return 0;
+	}
+
+	if (ZEND_TYPE_IS_CLASS(fe_type)) {
+		zend_string *fe_class_name =
+			_resolve_parent_and_self(fe, ZEND_TYPE_NAME(fe_type));
+		int code = 1;
+		if (ZEND_TYPE_IS_CLASS(proto_type)) {
+			zend_string *proto_class_name =
+				_resolve_parent_and_self(proto, ZEND_TYPE_NAME(proto_type));
 
-		proto_class_name = ZEND_TYPE_NAME(proto_arg_info->type);
-		class_name = ZSTR_VAL(proto_class_name);
-		class_name_len = ZSTR_LEN(proto_class_name);
-		if (class_name_len == sizeof("parent")-1 && !strcasecmp(class_name, "parent") && proto->common.scope && proto->common.scope->parent) {
-			proto_class_name = zend_string_copy(proto->common.scope->parent->name);
-		} else if (class_name_len == sizeof("self")-1 && !strcasecmp(class_name, "self") && proto->common.scope) {
-			proto_class_name = zend_string_copy(proto->common.scope->name);
+			if (!zend_string_equals_ci(fe_class_name, proto_class_name)) {
+				if (fe->common.type == ZEND_USER_FUNCTION) {
+					zend_class_entry * fe_ce = zend_lookup_class(fe_class_name);
+					zend_class_entry * proto_ce = zend_lookup_class(proto_class_name);
+
+					if (fe_ce && proto_ce) {
+						code = _check_zce_variance(fe_ce, proto_ce, variance);
+					} else {
+						/* The -1 will still be considered "truthy". It means there
+						 * is not enough information at the moment, but there are
+						 * not any known errors either. */
+						code = -1;
+					}
+				} else {
+					code = 0;
+				}
+			}
+			zend_string_release(proto_class_name);
+		} else if (variance == COVARIANT) {
+			if (proto_type_code == IS_ITERABLE) {
+				zend_class_entry * fe_ce = zend_lookup_class(fe_class_name);
+				code = fe_ce ? instanceof_function(fe_ce, zend_ce_traversable) : -1;
+			} else if (proto_type_code != IS_OBJECT) {
```

Comment:
I believe this needs to load the class as well, to be forward-compatible with `Foo`-style types that are not classes (but for example enums).

Reply:
I believe [zend_lookup_class](https://php-lxr.adamharvey.name/source/xref/master/Zend/zend_execute_API.c#zend_lookup_class) will autoload.

Reply:
Yes, it will. However you are not looking up any classes in this particular branch.